src/Orleans.BroadcastChannel/BroadcastChannelConsumerExtension.cs

```diff
@@ -0,0 +1,99 @@
+    {
+        private readonly ConcurrentDictionary<InternalChannelId, ICallback> _handlers = new();
+        private readonly IOnBroadcastChannelSubscribed _subscriptionObserver;
+        private AsyncLock _lock = new AsyncLock();
+
+        private interface ICallback
+        {
+            Task OnError(Exception exception);
+
+            Task OnPublished(object item);
+        }
+
+        private class Callback<T> : ICallback
+        {
+            private readonly Func<T, Task> _onPublished;
+            private readonly Func<Exception, Task> _onError;
+
+            private static Task NoOp(Exception _) => Task.CompletedTask;
+
+            public Callback(Func<T, Task> onPublished, Func<Exception, Task> onError)
+            {
+                _onPublished = onPublished;
+                _onError = onError ?? NoOp;
+            }
+
+            public Task OnError(Exception exception) => _onError(exception);
+
+            public Task OnPublished(object item)
+            {
+                return item is T typedItem
+                    ? _onPublished(typedItem)
+                    : _onError(new InvalidCastException($"Received an item of type {item.GetType().Name}, expected {typeof(T).FullName}"));
+            }
+        }
+
+        public BroadcastChannelConsumerExtension(IGrainContextAccessor grainContextAccessor)
+        {
+            _subscriptionObserver = grainContextAccessor.GrainContext?.GrainInstance as IOnBroadcastChannelSubscribed;
+        }
+
+        public async Task OnError(InternalChannelId streamId, Exception exception)
+        {
+            var callback = await GetStreamCallback(streamId);
+            if (callback != default)
+            {
+                await callback.OnError(exception);
+            }
+        }
+
+        public async Task OnPublished(InternalChannelId streamId, object item)
+        {
+            var callback = await GetStreamCallback(streamId);
+            if (callback != default)
+            {
+                await callback.OnPublished(item);
+            }
+        }
+
+        public void Attach<T>(InternalChannelId streamId, Func<T, Task> onPublished, Func<Exception, Task> onError)
+        {
+            _handlers.TryAdd(streamId, new Callback<T>(onPublished, onError));
+        }
+
+        private async ValueTask<ICallback> GetStreamCallback(InternalChannelId streamId)
+        {
+            ICallback callback;
+            if (_handlers.TryGetValue(streamId, out callback))
+            {
+                return callback;
+            }
+            using (await _lock.LockAsync())
+            {
+                if (_handlers.TryGetValue(streamId, out callback))
+                {
+                    return callback;
+                }
+                var subscription = new BroadcastChannelSubscription(this, streamId);
+                await _subscriptionObserver.OnSubscribed(subscription);
```

Comment:
The reason this works (`BroadcastChannelSubscription.OnSubscribed` calls `Attach`, which adds a handler) is not obvious from looking at the code - a descriptive comment would help